server/src/main/java/io/crate/role/TransportPrivilegesAction.java

```diff
@@ -161,5 +171,33 @@ static ApplyPrivsResult applyPrivileges(Metadata.Builder mdBuilder, PrivilegesRe
         return new ApplyPrivsResult(affectedRows, unknownRoleNames);
     }
 
+    private static void validateIsNotUser(Roles roles, RolePrivilegeToApply rolePrivilegeToApply) {
+        for (String roleNameToApply : rolePrivilegeToApply.roleNames()) {
+            if (roles.findRole(roleNameToApply).isUser()) {
+                throw new IllegalArgumentException("Cannot " + rolePrivilegeToApply.state().name() + " a USER to a ROLE");
+            }
+        }
+    }
+
+    @VisibleForTesting
+    static void detectCyclesInRolesHierarchy(Roles roles, PrivilegesRequest request) {
+        if (request.rolePrivilege().state() == PrivilegeState.GRANT) {
+            for (var grantee : request.roleNames()) {
+                for (var roleNameToGrant : request.rolePrivilege().roleNames()) {
+                    if (roleNameToGrant.equals(grantee)) {
+                        throw new IllegalArgumentException(String.format(Locale.ENGLISH,
+                            "Cannot grant role: %s to itself as a cycle will be created", grantee));
+                    }
+                    Set<String> parentsOfRoleToGrant = roles.findAllParents(roleNameToGrant);
```

Comment:
Looks like this can be moved outside of the loop as `parentsOfRoleToGrant` is equal for all `grantee` values.

Reply:
yep, I have to invert the 2 loops, great, thx!